### DISCUSSION TITLE
How to support negated `where` expression/condition

### DISCUSSION THREAD

**Discussion by brsanthu** [eclipse-langium/langium#1482] [Q&A] [answered]
We have been working on a custom SQL DSL that is based off of SQL. That DSL should support negated `where` expression (not just `not in` operator) and conditions. Below are some examples that it should support and here is the simplified part [playground](https://langium.org/playground?grammar=OYJwhgthYgBAwgZwI4BsBQ6CmA7ALiAJ4IqoCyA9gCZaqwCWOAZliIiWpTagFzqwDYiPGDxYIuPIgDUAXgDKIsRPwAqdAG5Mi0eMl9BHOgHINxgPybMSNAcHHEtLAGM8x-oKb1aVGbIBi3qhUAgAUxgA0xrBePn6BPgIAlOqGxkwgFBDRFABGAFYueLIAcpBYHgIA7gAWrFiyAOp1IFjwqGAAro6WWujN9e1djnYCxrX10VgAHgAOIE0tWACicyBW-Uur86OwAPIg2%2Bt9B0cMzKzsA61HuwCCOFRnoQDe5yxwp2sAdKhYTMVnJ0QK18ABfWDGCggaIgejAGrFB5PNbJdR9ZFnRgfK5bNa7AAKcOgRGeb2xrFgmJ%2BfwBsiBIMkEOMYEesPhiNkRPoJMIZ1gKQ23N5WIubFg1xW%2BMqsFmMFw5lkxnCErx81l8pwiuMSWisAAPqrBhRHhtJfATVRdrTiglgrBQhRZqwShQ8NqcG6LElYE7WEqlQI4QjipKAGpgVCdCp9cOR6O7RRwnDACNRmPaAiMVPxrC7ABuuYUAEkSgBxAAyywA%2BvIACoAJVLZY2dpCuxw5VK5Q2ZQksFaeGBOHYwmTwF2UJANBhBtgxYAIhsxCAIIxI-OFzxYAB6ADa1bAAFoAF53I8ALQAugfj2fLwAGI8ATivqh3WhXa87dHkzartaNs2A5YEOIAjkIWYpjwABExgwQ6xgADpIdE3xzrAACE4QodEhpwQhSQpARAh9DU9BUDQOCwF%2B650GQFbVvAexkGQywlHW247kh3GqHuSGIEh8hvuYSGqDxH7oORlG4DRrDfhu8iMcxrHsZxu48Txe4AHpITgSEgG%2BklSRRVFyaudESvIIFgRBY7ZjwKp4ZCSFuHOyGzoayE4NESTSFoQA&content=PTAEHUEsBcAtQHIHtqgMJIHYBMaSwM4BQBApgDakDGqk2ANKJgIYC2poAZgE5KuhIARgCtq0IgHdYpbh0wpQdUAF5QAcmikC0NQG4iJCmMUMmbDjz4CRYydNmgAFPNrYV6zdrUBKfYco0Jows7Fy8-EKiNHYyHM4KjkqqGlo6oMw4ZqHJnjrevgZkAa7B5mFWkbZSsU4uTkkeqWrpmSEcOU35fiAQMLBIAK6oyKgYOHiE-sZ0paGWETbR1Q4NKV5%2BRdOmbeULUeLLcau5PhtGgTNZFuHW%2BzErbh1eLW47TzpnxUFXu7dV9nFEo9Gl5vC9amV3j4ClMLtsyvM-ksAU4ge41nlwc5ISC8mCkNx6sCMQAmU6Fc4lH6IyrImqONF1KFgjJubHZXHQgSEtHvMkFIA).

```
where not id = 'test';

where (not id = 'test');

where (not (id = 'test' and name = 'test'));

where (not (id = 'test' and name = 'test')) or (name = 'test');
```

I have tried many iterations and I'm not getting where to place that negated expr grammar. We went through the [langium-sql](https://github.com/TypeFox/langium-sql/blob/c20515756c30779a34c73d81238da8ec94cc1ab3/packages/langium-sql/src/sql.langium#L158-L159) where clause but still not getting how to support this type of construct. So far EBNF has been very easy to follow except for where expressions due to its recursive nature.

**Comment by msujew** [comment, marked as answer]
Hey @brsanthu,

this is effectively just another type of primary expression, see [this modified playground example](https://langium.org/playground?grammar=OYJwhgthYgBAwgZwI4BsBQ6CmA7ALiAJ4IqoCyA9gCZaqwCWOAZliIiWpTagFzqwDYiPGDxYIuPIgDUAXgDKIsRPwAqdAG5Mi0eMl9BHOgHINxgPybMSNAcHHEtLAGM8x-oKb1aVGbIBi3qhUAgAUxgA0xrBePn6BPgIAlOqGxkwgFBDRFABGAFYueLIAcpBYHgIA7gAWrFiyAOp1IFjwqGAAro6WWujN9e1djnYCxrX10VgAHgAOIE0tWACicyBW-Uur8wzMrOwDrdsgo7AA8iDHGxfHuyxssIcra6cAgjhUt6EA3nes55c1gA6VBYJjFZydECtfAAX1gxgoIGiIHowBqxXenzWyXUfSxt0Y9wOWxelVgAAVUdAiF9fkT-gTgaDwbJIdDJPDjGAPii0RjZFT6DTCLdYCkNkKRYS9g8nsdTsZwo9STtjElogAfFWDCgfWDakoUPBXPpGk04hly1UnWDfK2wc3HLk4Y1TNaLerHPpPeB6qinFnFBLBWChCizVjm8yyYyutzmJKwCOsWOxgSo9HFJ4ANTAqE6FR9SzzBawp0UqJwwFLhY2lcYNfzhdOADdmw15ABJEoAcQAMssAPryAAqACUe72NiGQqccOVSuUNmUJLBWngoTh2MIq8BFUiaMiDbAuwARDZiEAQRj509nniwAD0AG0h2AALQAL1eH4AWgAum%2Bn4-v%2BAAMH4AJwAaoT5aFeN4LnQ3Z9oOI4TlO65YJuIDbkIBCNjwABExhEWGxgADoUdEQInrAACE4RUVqsAkWRSQpGxAh9DU9BUDQOCwAht50GQ-ZDvAZxkGQywlKOj5PhRimqC%2BFGIBR8gweYFGqEpcHoLx-G4EJrCIXe8jiZJ0myfJz5KUpL4AHoUTgFEgDB%2BkGXxAkmdeImPPIWE4Xhu6EcqLGUW4J6Uce2qUTg0RJNIWhAA&content=PTAEHUEsBcAtQHIHtqgMJIHYBMaSwM4BQBApgDakDGqk2ANKJgIYC2poAZgE5KuhIARgCtq0IgHdYpbh0wpQdUAF5QAcmikC0NQG4iJCmMUMmbDjz4CRYydNmgAFPNrYV6zdrUBKfYco0Jows7Fy8-EKiNHYyHM4KjkqqGlo6oMw4ZqHJnjrevgZkAa7B5mFWkbZSsU4uTkkeqWrpmSEcOU35fiAQMLBIAK6oyKgYOHiE-sZ0paGWETbR1Q4NKV5%2BRdOmbeULUeLLcau5PhtGgTNZFuHW%2BzErbh1eLW47TzpnxUFXu7dV9nFEo9Gl5vC9amV3j4ClMLtsyvM-ksAU4ge41nlwc5ISC8mCkNx6sCMQAmU6Fc4lH6IyrImqONF1KFgjJubHZXHQgSEtHvMkFIA).

**Reply by brsanthu** [reply]
Thank you, that looks so elegant after we pulling our hair to get it work :).

Can you please explain what does `{infer NotExpr}` means?

**Reply by cdietrich** [reply]
it basically means create an instance of `NotExpr` and infer its type.
see https://langium.org/docs/grammar-language/#simple-actions
